### PR TITLE
added more exception handling

### PR DESCRIPTION
**Pull request by harsha-simhadri**

### CODE REVIEW

File: include/ann_exception.h [hunk 1]
```diff
@@ -3,28 +3,34 @@
...
 namespace diskann {
-  class ANNException {
+
+  class ANNException : public std::runtime_error {
    public:
     DISKANN_DLLEXPORT ANNException(const std::string& message, int errorCode);
     DISKANN_DLLEXPORT ANNException(const std::string& message, int errorCode,
                                    const std::string& funcSig,
                                    const std::string& fileName,
                                    unsigned int       lineNum);
```
Comment (gopalrs) [inline include/ann_exception.h:23]: should we override "what()" here?

File: include/ann_exception.h [hunk 2]
```diff
@@ -3,28 +3,34 @@
...
+    int _errorCode;
+  };
+
+  class FileException : public ANNException {
+   public:
+    DISKANN_DLLEXPORT FileException(const std::string& filename,
+                                    std::system_error& e,
+                                    const std::string& funcSig,
+                                    const std::string& fileName,
+                                    unsigned int       lineNum);
```
Comment (gopalrs) [inline include/ann_exception.h:34]: Here too? (Override the what() function)

File: include/cached_io.h [hunk 3]
```diff
@@ -99,12 +107,17 @@ class cached_ofstream {
...
   cached_ofstream(const std::string& filename, uint64_t cache_size)
       : cache_size(cache_size), cur_off(0) {
-    writer.open(filename, std::ios::binary);
-    assert(writer.is_open());
-    assert(cache_size > 0);
-    cache_buf = new char[cache_size];
-    diskann::cout << "Opened: " << filename.c_str()
-                  << ", cache_size: " << cache_size << std::endl;
+    try {
+      writer.open(filename, std::ios::binary);
```
Comment (gopalrs) [inline include/cached_io.h:111]: Does writer.open() throw exceptions as default behavior ?

**Review by gopalrs — APPROVED** [review]

File: include/cached_io.h (hunk 3 above)
Comment (harsha-simhadri) [reply include/cached_io.h:112]: I added exception flags here.

**Pull request merged**